File: LeetCode/Medium/0017. Letter Combinations of a Phone Number.java
class Solution {
    // set up hashmap
    static HashMap<String, List<String>> hmap = new HashMap();
    static boolean flag = false;

    public void fillHashMap() {
        for(int i = 2; i <= 9; ++i) {
            Integer id = new Integer(i);
            ArrayList<String> l = new ArrayList<String>();

            switch(id.toString()) {
                case "2":
                    l.add("a"); l.add("b"); l.add("c"); break;
                case "3":
                    l.add("d"); l.add("e"); l.add("f"); break;
                case "4":
                    l.add("g"); l.add("h"); l.add("i"); break;
                case "5":
                    l.add("j"); l.add("k"); l.add("l"); break;
                case "6":
                    l.add("m"); l.add("n"); l.add("o"); break;
                case "7":
                    l.add("p"); l.add("q"); l.add("r"); l.add("s"); break;
                case "8":
                    l.add("t"); l.add("u"); l.add("v"); break;
                case "9":
                    l.add("w"); l.add("x"); l.add("y"); l.add("z"); break; 
            }

            hmap.put(id.toString(), l);
        }
    }
        

    public List<String> letterCombinations(String digits) {
        if(flag == false) {
            fillHashMap();
            flag = true;
        }
        
        if(digits.length() == 0) return new ArrayList<String>();
        else if(digits.length() == 1) return hmap.get(digits);
        
        List<String> list = new ArrayList<String>();
        List<String> oldList = letterCombinations(digits.substring(1));

        for(String letter : hmap.get(digits.substring(0, 1))) {
            for(String letterComb : oldList) {
                list.add(letter + letterComb);
            }
        }

        return list;
    }
}

File: LeetCode/Medium/0017. Letter Combinations of a Phone Number.py
phone_map = {
    "2": ["a", "b", "c"],
    "3": ["d", "e", "f"],
    "4": ["g", "h", "i"],
    "5": ["j", "k", "l"],
    "6": ["m", "n", "o"],
    "7": ["p", "q", "r", "s"],
    "8": ["t", "u", "v"],
    "9": ["w", "x", "y", "z"]
}

class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]

        idea:
         - letterCombinations(digits[1:]) returns all comb for digits[1:]
         - then add all possible letter for digits[0]
        """
        
        if len(digits) == 0: return []
        elif len(digits) == 1: return phone_map[digits]

        ans = []
        remaining_comb = self.letterCombinations(digits[1:])

        for letter in phone_map[digits[0]]:
            for comb in remaining_comb:
                ans.append(letter + comb)
        
        return ans


        
